cloudinit/sources/helpers/azure.py

```diff
@@ -777,13 +809,28 @@ def register_with_azure_and_fetch_data(self, pubkey_info=None) -> dict:
         health_reporter.send_ready_signal()
         return {'public-keys': ssh_keys}
 
+    @azure_ds_telemetry_reporter
+    def register_with_azure_and_report_failure(self, description: str) -> None:
+        """Gets the VM's GoalState from Azure, uses the GoalState information
+        to report failure/send provisioning failure signal to Azure.
+
+        @param: user visible error description of provisioning failure.
+        """
+        if self.azure_endpoint_client is None:
+            self.azure_endpoint_client = AzureEndpointHttpClient(None)
```

Comment:
This instantiation of `self.azure_endpoint_client` differs from the instantiation we already perform; it will never pass in a certificate, whereas the other instantiation sometimes will. I'm concerned that this inconsistency could lead to issues: depending on which of the methods is called first, `self.azure_endpoint_client` will behave differently.

Reply:
Thanks for pointing it out @OddBloke.

I think that situation would only happen if the same `WALinuxAgentShim` instance invokes both report ready and report failure. The report ready and report failure codepaths are separate, and the corresponding `WALinuxAgentShim` for both codepaths are instantiated separately [(here)](https://github.com/johnsonshi/cloud-init/blob/07ac8bfc2d088bb568b5010b4f8b470e15e77fa8/cloudinit/sources/helpers/azure.py#L962-L984)

I don't think we'll ever hit the situation where the same `WALinuxAgentShim` instance invokes both report ready and report failure.